modules/index/src/main/java/org/apache/ignite/internal/index/IndexManager.java

```diff
@@ -198,6 +195,48 @@ private CompletableFuture<Boolean> onIndexCreate(CreateIndexEventParameters para
         });
     }
 
+    private void onIndexRemoved(RemoveIndexEventParameters parameters) {
+        inBusyLock(busyLock, () -> {
+            int indexId = parameters.indexId();
+            int catalogVersion = parameters.catalogVersion();
+            int previousCatalogVersion = catalogVersion - 1;
+
+            // Retrieve descriptor during synchronous call, before the previous catalog version could be concurrently compacted.
+            CatalogIndexDescriptor indexDescriptor = catalogService.index(indexId, previousCatalogVersion);
+            assert indexDescriptor != null : "indexId=" + indexId + ", catalogVersion=" + previousCatalogVersion;
+
+            int tableId = indexDescriptor.tableId();
+
+            if (catalogService.table(tableId, catalogVersion) == null) {
+                // Nothing to do. Index will be destroyed along with the table.
+                return;
+            }
+
+            destructionEventsQueue.enqueue(new DestroyIndexEvent(catalogVersion, indexId, tableId));
```

Comment:
It seems there may be a race, if `onLwmChanged` is executed in parallel, we may skip the index which could be started to be destroyed immediately, we will have to wait for the next lwm update. This is not scary if the sign is updated every second, but if it’s a day, then we can keep the data on disk for quite a long time.

We need to somehow fix this race so as not to delay the metastor thread.

Reply:
This code is called for a the new CatalogVersion, which is not ready yet.
An index is `alive` in the actual Catalog version (relative to LWM timestamp), so it will not be cleaned by LWM listener. 
Thus we can't call onLwmChange synchronously here not only because of Metastorage thread.

Reply:
https://issues.apache.org/jira/browse/IGNITE-21755